### PR TITLE
feat: override rules

### LINKED ISSUE #171: Defined root paths / quality profiles on a per-user basis
### Description

Currently with Ombi I'm able to set the default root path and quality/language profile on a per user basis. This allows for people who are not technically inclined to not have to worry about selecting a root folder or quality setting (and therefore I don't have to check they did it right every time they make a request). 

### Desired Behavior

Something like Ombi where quality profiles and root paths can be set both globally, but also on a per user basis. 

image

### Additional Context

_No response_

### Code of Conduct

- [X] I agree to follow Overseerr's Code of Conduct

### LINKED ISSUE #419: [Feature Request] Possibility to specify Sonarr/Radarr root folder per user
### Description

I'd love to specify different root folders for different users.
Example:
Sonarr is configured with following root folders:
`/media/user1`
`/media/user2`
`/media/user3`

With Jellyfin, user1 has access to their personal library which is located at `/media/user1`.
The same goes for Jellyfin accounts user2 -> `/media/user2` and user3 -> `/media/user3`

Using Jellyseerr, there is indeed the possibility to grant users the permission `Advanced Requests` so they can manually choose the root folder where they want to have their downloaded content.

But sometimes - especially if those root folders are meant to be user-specific personal Jellyfin libraries - an option to specify one specific root folder would be very nice.
Because even if the users are aware that they have to change this value - eventually they might forget about it.

### Desired Behavior

One possibility to achieve that:
Add a possibility under _Settings_ -> _Services_ -> _Edit Radar/Sonarr_ -> _Root Folder_ to optionally add different root folders linked to one/several Jellyseerr users.

Another place where such an option could be integrated is, of course, _Settings_ -> _Users_, where a submenu could be added with the option to link Sonarr/Radarr root folders to one/several Jellyseerr users.

The same goes for _Users_ -> _user1_ -> _View Profile_ -> _Permissions_

I don't know how hard it is to implement such an option but one similar idea is probably #197

Thanks for all your hard work! 

### Additional Context

_No response_

### Code of Conduct

- [X] I agree to follow Overseerr's Code of Conduct

### LINKED ISSUE #909: More root path besides normal and anime
### Description

At the Moment only normal tv shows, movies and anime shows get the right root path. For anime movies or documantarys for example you have to do it manually. 

### Desired Behavior

It would be nice if jellyseer could Auto path documantary for example to the right root path

### Additional Context

_No response_

### Code of Conduct

- [X] I agree to follow Jellyseerr's Code of Conduct

### LINKED ISSUE #921: Different quality profiles dor different users
### Description

I want specific users to have their quality profile set for 1080p and some to have a 4k quality profile. This should be as simple as saving in DB for each user a specific quality profile.

### Desired Behavior

When changing users permissions give them the option to use a default quality profile that differs from user to user

### Additional Context

I run everything on a RPI, which is good enough for direct play but it is a nogo for transcode. I want to make this available to my family but most of em don't have 4k compatible devices and don't have enough understanding to select their own quality profile. So it would be ideal so i can set one for each one of them.

### Code of Conduct

- [X] I agree to follow Jellyseerr's Code of Conduct

### PR DESCRIPTION
**Pull request by gauthier-th**
#### Description

This PR introduces override rules in Jellyseerr.
Override rules are used to define conditions for each service (i.e. for each Radarr/Sonarr server) on which parameters will be modified.
For example, it's possible to define a different quality profile for certain users, or change the root path according to the genre of a film or series, etc.

#### Screenshot (if UI-related)

image
image

#### To-Dos

- [x] Successful build `pnpm build`
- [x] Translation keys `pnpm i18n:extract`
- [ ] Database migration (if required)

#### Issues Fixed or Closed

- Fixes #171
- Fixes #419
- Fixes #909
- Fixes #921

### CODE REVIEW

**Comment by juancarlosromerogarcia** [comment]
lint & Test Build, Post Setup pnpm cache,
`/bin/tar --posix -cf cache.tgz --exclude cache.tgz -P -C /__w/jellyseerr/jellyseerr --files-from manifest.txt -z`
/bin/tar: unrecognized option: posix

**Comment by gauthier-th** [comment]
> lint & Test Build, Post Setup pnpm cache,
> `/bin/tar --posix -cf cache.tgz --exclude cache.tgz -P -C /__w/jellyseerr/jellyseerr --files-from manifest.txt -z`
> /bin/tar: unrecognized option: posix

What is this supposed to mean?

**Comment by gauthier-th** [comment]
For anyone willing to test this, it's available with the preview tag: `preview-override-rules`
PS: remember to do backups :wink:

**Comment by feerlessleadr** [comment]
@gauthier-th I just pulled this image and am doing some testing. For some reason when I try to add an override rule to either sonarr or radarr, I'm not able to create the rule. When I click on 'Creat Rule' nothing happens, and I don't see anything (either errors or anything at all) either the jellyseer log or the docker logs. Here is the rule I'm trying to create for sonarr:

image

**Comment by gauthier-th** [comment]
> @gauthier-th I just pulled this image and am doing some testing. For some reason when I try to add an override rule to either sonarr or radarr, I'm not able to create the rule. When I click on 'Creat Rule' nothing happens, and I don't see anything (either errors or anything at all) either the jellyseer log or the docker logs. Here is the rule I'm trying to create for sonarr:
> 
> image

I just rebased the PR to the latest commit of develop, can you please test again with this version? (you'll need to update the Docker image)

If it's still not working, could you check Jellyseerr logs as well as your browser logs and see if there is anything there? (F12 -> Console)

**Comment by Fallenbagel** [comment]
> I have tried it but nothing happens when I press the `create rule` button. In my browser console, every so often, it shows a 404 error: message "`SQLITE_ERROR: no such table: override_rule`"

That means your migrations didn't apply. Can you test with a fresh config first?

**Comment by starlingfire** [comment]
I just tried testing this as well, and can't get the override screen to appear. Here's how I set it up, in case I did something wrong:

1. Cloned this repo
2. Changed to develop branch: `git branch develop`
3. Build/run with docker compose: `docker compose up -d`
4. Connect to local Radarr and Jellyfin instances

My Radarr edit screen looks like this:

image

The About page shows I'm on develop:

image

Ran the following to confirm my working environment:
```
root@playground:~/jellyseerr# git switch develop
Already on 'develop'
Your branch is up to date with 'origin/develop'.
```

**Comment by Fallenbagel** [comment]
> I just tried testing this as well, and can't get the override screen to appear. Here's how I set it up, in case I did something wrong:
> 
> 1. Cloned this repo
> 2. Changed to develop branch: `git branch develop`
> 3. Build/run with docker compose: `docker compose up -d`
> 4. Connect to local Radarr and Jellyfin instances
> 
> My Radarr edit screen looks like this:
> 
> image
> 
> The About page shows I'm on develop:
> 
> image
> 
> Ran the following to confirm my working environment:
> ```
> root@playground:~/jellyseerr# git switch develop
> Already on 'develop'
> Your branch is up to date with 'origin/develop'.
> ```

It's not in develop branch. It's in `override rules`. Also you don't need to build the image we already have `preview-override-rules` as an image...

**Comment by starlingfire** [comment]
> It's not in develop branch. It's in `override rules`. Also you don't need to build the image we already have `preview-override-rules` as an image...

Thank you!

Got the preview image running, and I get 404 errors in the browser log trying to save the override rule (and on a recurring basis simply while on that screen).
```
XHRGET
http://192.168.1.205:5055/api/v1/overrideRule
[HTTP/1.1 404 Not Found 8ms]

	
GET
	http://192.168.1.205:5055/api/v1/overrideRule
Status
404
Not Found
VersionHTTP/1.1
Transferred298 B (56 B size)
Referrer Policystrict-origin-when-cross-origin
DNS ResolutionSystem

    	
    Connection
    	keep-alive
    Content-Length
    	56
    Content-Type
    	application/json; charset=utf-8
    Date
    	Mon, 18 Nov 2024 18:38:16 GMT
    ETag
    	W/"38-[redacted]"
    Keep-Alive
    	timeout=5
    X-Powered-By
    	Express
    	
    Accept
    	*/*
    Accept-Encoding
    	gzip, deflate
    Accept-Language
    	en-US,en;q=0.5
    Connection
    	keep-alive
    Cookie
    	connect.sid=[redacted]
    DNT
    	1
    Host
    	192.168.1.205:5055
    Priority
    	u=4
    Referer
    	http://192.168.1.205:5055/settings/services
    Sec-GPC
    	1
    User-Agent
    	Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0
```

**Comment by gauthier-th** [comment]
Oops, I may have forgotten to write the migration. I'll write them and it should work better 😅

**Comment by gauthier-th** [comment]
@starlingfire the migration script has been added. Can you please test again? (you'll need to update the Docker image)

**Comment by starlingfire** [comment]
@gauthier-th so far, so good. I added an override rule to put a user's downloads in a different root folder, and it appears to be working. I'll mess around with it more tomorrow when I get a bit more time. I'm so excited to have this feature!

**Comment by feerlessleadr** [comment]
Just tested on my system and it seems to be working as well. Thanks for this!

**Comment by juancarlosromerogarcia** [comment]
I have tested both sonarr and radarr with various rules that assign root folders depending on the user. It works without problems.
[gauthier-th](https://github.com/gauthier-th) thanks a lot for your work.

**Comment by Yankees4life** [comment]
@Fallenbagel 

I hope this gets merged. It will certainly help me set up Radarr Anime instances easily

File: server/migration/1731963944025-AddOverrideRules.ts
Comment (fallenbagel) [inline server/migration/1731963944025-AddOverrideRules.ts:1]: This needs to be moved into sqlite folder and a postgres migration needs to be generated right?

**Review by Fallenbagel — CHANGES_REQUESTED** [review]

File: server/migration/1731963944025-AddOverrideRules.ts
Comment (gauthier-th) [reply server/migration/1731963944025-AddOverrideRules.ts:1]: Yep

**Comment by actuallyevan** [comment]
Would this allow routing to multiple *arr instances simultaneously? Really looking forward to this being merged!

**Comment by gauthier-th** [comment]
> Would this allow routing to multiple *arr instances simultaneously?

No, this will not change the behavior of the *arr instances, just change which parameters are passed to the instance when a media is requested.

**Comment by starlingfire** [comment]
> Would this allow routing to multiple *arr instances simultaneously? Really looking forward to this being merged!

This _will_ let you route media to different root folders based on who made the request, so each user can add to their own individual library. But you can do that within the same instance of the arr stack, you don't need multiple.

**Comment by actuallyevan** [comment]
> > Would this allow routing to multiple *arr instances simultaneously? Really looking forward to this being merged!
> 
> This _will_ let you route media to different root folders based on who made the request, so each user can add to their own individual library. But you can do that within the same instance of the arr stack, you don't need multiple.

Yeah, I figured. Eventually hoping for something like [this](https://github.com/varthe/Redirecterr) to become native functionality in a *seerr

**Comment by Dental1347** [comment]
> > Would this allow routing to multiple *arr instances simultaneously?
> 
> No, this will not change the behavior of the *arr instances, just change which parameters are passed to the instance when a media is requested.

I've got a question with a usecase.
Currently, it's near impossible to setup a sonarr instance for series and a sonarr instance for anime. Because, the series tab is require to be filled in.

I guess that as a workaround, it could be possible to setup the sonarr series, and not configure anime, and set this as default. But, for the sonarr anime, you configure series as a dummy, and setup anime as you would normally.

With this pull, is it then possible to setup a condition that:
1. if it's anime, send it to the sonarr anime instance, instead of the sonarr series?

**Comment by gauthier-th** [comment]
>  Currently, it's near impossible to setup a sonarr instance for series and a sonarr instance for anime. Because, the series tab is require to be filled in.

This is something #232 should solve, and it's on our roadmap.

File: src/components/Discover/DiscoverTvUpcoming.tsx [hunk 1]
```diff
@@ -7,7 +7,9 @@ import defineMessages from '@app/utils/defineMessages';
 import type { TvResult } from '@server/models/Search';
 import { useIntl } from 'react-intl';
 
-const messages = defineMessages('components.DiscoverTvUpcoming', {});
+const messages = defineMessages('components.DiscoverTvUpcoming', {
+  upcomingtv: 'Upcoming Series',
```
Comment (fallenbagel) [inline src/components/Discover/DiscoverTvUpcoming.tsx]: Is this relevant for this PR?

File: src/components/Settings/SonarrModal/index.tsx [hunk 2]
```diff
@@ -75,48 +82,56 @@ const messages = defineMessages('components.Settings.SonarrModal', {
...
+}: SonarrModalProps) => {
   const intl = useIntl();
+  const { data: rules, mutate: revalidate } =
+    useSWR<OverrideRuleResultsResponse>('/api/v1/overrideRule');
   const initialLoad = useRef(false);
   const { addToast } = useToasts();
   const [isValidated, setIsValidated] = useState(sonarr ? true : false);
   const [isTesting, setIsTesting] = useState(false);
-  const [testResponse, setTestResponse] = useState<TestResponse>({
+  const [testResponse, setTestResponse] = useState<SonarrTestResponse>({
```
Comment (fallenbagel) [inline src/components/Settings/SonarrModal/index.tsx:122]: Why is there a difference in consistency for test response in sonarr modal compared to radarr modal

File: src/utils/defineMessages.ts
Comment (fallenbagel) [inline src/utils/defineMessages.ts:1]: Is this change relevant to this PR?

**Comment by actuallyevan** [comment]
> > > Would this allow routing to multiple *arr instances simultaneously?
> > 
> > 
> > No, this will not change the behavior of the *arr instances, just change which parameters are passed to the instance when a media is requested.
> 
> I've got a question with a usecase. Currently, it's near impossible to setup a sonarr instance for series and a sonarr instance for anime. Because, the series tab is require to be filled in.
> 
> I guess that as a workaround, it could be possible to setup the sonarr series, and not configure anime, and set this as default. But, for the sonarr anime, you configure series as a dummy, and setup anime as you would normally.
> 
> With this pull, is it then possible to setup a condition that:
> 
> 1. if it's anime, send it to the sonarr anime instance, instead of the sonarr series?

I would use [Redirecterr](https://github.com/varthe/Redirecterr). There is also a custom Overseerr (not jelly) anime instance floating around. This is how I do it, and I think achieves the functionality you're looking for.

**Review by Fallenbagel — APPROVED** [review]

**Comment by jerry00020001** [comment]
Would it be possible to add an override rule based on date released?

**Comment by gauthier-th** [comment]
> Would it be possible to add an override rule based on date released?

Please open an issue for a feature request. This is not the right place to ask.

**Pull request merged**